src/audio/google/google_rtc_audio_processing.c

```diff
@@ -640,125 +591,191 @@ static int google_rtc_audio_processing_reset(struct processing_module *mod)
 		}
-		num_samples_remaining -= n;
-		ref = audio_stream_wrap(ref_stream, ref);
+		ref += cd->num_aec_reference_channels;
+		if ((void *)ref >= (void *)ref_buf_end)
+			ref = (void *)ref_buf_start;
 	}
-	input_buffers[cd->aec_reference_source].consumed = num_aec_reference_bytes;
-
-	mic_streamb = &input_buffers[cd->raw_microphone_source];
-	mic_stream = mic_streamb->data;
-	out_streamb = &output_buffers[0];
-	out_stream = out_streamb->data;
-
-	src = audio_stream_get_rptr(mic_stream);
-	dst = audio_stream_get_wptr(out_stream);
-
-	frames = input_buffers[cd->raw_microphone_source].size;
-	num_frames_remaining = frames;
-
-	while (num_frames_remaining) {
-		nmax = audio_stream_frames_without_wrap(mic_stream, src);
-		n = MIN(num_frames_remaining, nmax);
-		nmax = audio_stream_frames_without_wrap(out_stream, dst);
-		n = MIN(n, nmax);
-		for (i = 0; i < n; i++) {
-			memcpy_s(&(cd->raw_mic_buffer[cd->raw_mic_buffer_frame_index *
-						      cd->num_capture_channels]),
-				 cd->num_frames * cd->num_capture_channels *
-				 sizeof(cd->raw_mic_buffer[0]), src,
-				 sizeof(int16_t) * cd->num_capture_channels);
-			++cd->raw_mic_buffer_frame_index;
-
-			memcpy_s(dst, cd->num_frames * cd->num_capture_channels *
-				 sizeof(cd->output_buffer[0]),
-				 &(cd->output_buffer[cd->output_buffer_frame_index *
-						     cd->num_capture_channels]),
-				 sizeof(int16_t) * cd->num_capture_channels);
-			++cd->output_buffer_frame_index;
-
-			if (cd->raw_mic_buffer_frame_index == cd->num_frames) {
-				GoogleRtcAudioProcessingProcessCapture_int16(cd->state,
-									     cd->raw_mic_buffer,
-									     cd->output_buffer);
-				cd->output_buffer_frame_index = 0;
-				cd->raw_mic_buffer_frame_index = 0;
-			}
 
-			src += audio_stream_get_channels(mic_stream);
-			dst += audio_stream_get_channels(out_stream);
-		}
-		num_frames_remaining -= n;
-		src = audio_stream_wrap(mic_stream, src);
-		dst = audio_stream_wrap(out_stream, dst);
+	GoogleRtcAudioProcessingAnalyzeRender_float32(
+			cd->state,
+			(const float **)cd->aec_reference_buffer_ptrs);
+
+	source_release_data(ref_stream, num_of_bytes_to_process);
+
+	/* process main stream - de interlace and convert */
+	num_of_bytes_to_process = cd->num_frames * source_get_frame_bytes(src_stream);
+	ret = source_get_data(src_stream, num_of_bytes_to_process, (const void **)&src,
+			      (const void **)&src_buf_start,  &src_buf_size);
+	assert(!ret);
+	src_buf_end = src_buf_start + src_buf_size;
+
+	for (int i = 0; i < cd->num_frames; i++) {
+		for (channel = 0; channel < cd->num_capture_channels; channel++)
+			cd->process_buffer_ptrs[channel][i] =
+					convert_uint16_to_float_hifi(src[channel]);
+
+		src += cd->num_capture_channels;
+		if ((void *)src >= (void *)src_buf_end)
+			src = (void *)src_buf_start;
+	}
+
+	source_release_data(src_stream, num_of_bytes_to_process);
+
+	/* call the library, use same in/out buffers */
+	GoogleRtcAudioProcessingProcessCapture_float32(cd->state,
```

Comment:
Are we sure the playback and the capture are in sync: ie they get 10ms of data at the same time ? I see nothing in the code ensuring it is the case: I might have missed it.
Otherwise we take care of that by calling the `AnalyseRender` only when 10ms of playback data, and only call the `ProcessCapture` when we have 10ms of raw microphone data.

Reply:
What if 10ms playback is ready 9ms after 10ms capture are ready ? That will leave a deadline of 1ms wouldn't it ?

Reply:
Not to mentioned the case when the rates are not exactly the same on render and capture (I think we can omit this case for now).